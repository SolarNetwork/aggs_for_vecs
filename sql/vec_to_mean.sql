-- smallint mean lots" {
SELECT vec_to_mean(smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- smallint mean none" {
SELECT vec_to_mean(smallints) FROM measurements WHERE sensor_id = -1;
-- smallint mean one null" {
SELECT vec_to_mean(smallints) FROM measurements WHERE sensor_id = 1;
-- smallint mean array of nulls" {
SELECT vec_to_mean(smallints) FROM measurements WHERE sensor_id = 2;
-- smallint mean one not-null" {
SELECT vec_to_mean(smallints) FROM measurements WHERE sensor_id = 4;
-- smallint mean array of nulls and one other" {
SELECT vec_to_mean(smallints) FROM measurements WHERE sensor_id IN (2, 4);

-- int mean lots" {
SELECT vec_to_mean(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- int mean none" {
SELECT vec_to_mean(ints) FROM measurements WHERE sensor_id = -1;
-- int mean one null" {
SELECT vec_to_mean(ints) FROM measurements WHERE sensor_id = 1;
-- int mean array of nulls" {
SELECT vec_to_mean(ints) FROM measurements WHERE sensor_id = 2;
-- int mean one not-null" {
SELECT vec_to_mean(ints) FROM measurements WHERE sensor_id = 4;
-- int mean array of nulls and one other" {
SELECT vec_to_mean(ints) FROM measurements WHERE sensor_id IN (2, 4);

-- bigint mean lots" {
SELECT vec_to_mean(bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- bigint mean none" {
SELECT vec_to_mean(bigints) FROM measurements WHERE sensor_id = -1;
-- bigint mean one null" {
SELECT vec_to_mean(bigints) FROM measurements WHERE sensor_id = 1;
-- bigint mean array of nulls" {
SELECT vec_to_mean(bigints) FROM measurements WHERE sensor_id = 2;
-- bigint mean one not-null" {
SELECT vec_to_mean(bigints) FROM measurements WHERE sensor_id = 4;
-- bigint mean array of nulls and one other" {
SELECT vec_to_mean(bigints) FROM measurements WHERE sensor_id IN (2, 4);

-- real mean lots" {
SELECT vec_to_mean(reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- real mean none" {
SELECT vec_to_mean(reals) FROM measurements WHERE sensor_id = -1;
-- real mean one null" {
SELECT vec_to_mean(reals) FROM measurements WHERE sensor_id = 1;
-- real mean array of nulls" {
SELECT vec_to_mean(reals) FROM measurements WHERE sensor_id = 2;
-- real mean one not-null" {
SELECT vec_to_mean(reals) FROM measurements WHERE sensor_id = 4;
-- real mean array of nulls and one other" {
SELECT vec_to_mean(reals) FROM measurements WHERE sensor_id IN (2, 4);

-- float8 mean lots" {
SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- float8 mean none" {
SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id = -1;
-- float8 mean one null" {
SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id = 1;
-- float8 mean array of nulls" {
SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id = 2;
-- float8 mean one not-null" {
SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id = 4;
-- float8 mean array of nulls and one other" {
SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id IN (2, 4);
