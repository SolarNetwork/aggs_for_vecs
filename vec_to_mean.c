
Datum vec_to_mean_transfn(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(vec_to_mean_transfn);

/**
 * Returns an of n elements,
 * which each element is the mean value found in that position
 * from all input arrays.
 *
 * by Paul A. Jungwirth
 */
Datum
vec_to_mean_transfn(PG_FUNCTION_ARGS)
{
  Oid elemTypeId;
  int16 elemTypeWidth;
  bool elemTypeByValue;
  char elemTypeAlignmentCode;
  int currentLength;
  MemoryContext aggContext;
  VecArrayBuildState *state = NULL;
  ArrayType *currentArray;
  int arrayLength;
  Datum *currentVals;
  bool *currentNulls;
  int i;

  if (!AggCheckCallContext(fcinfo, &aggContext)) {
    elog(ERROR, "vec_to_mean_transfn called in non-aggregate context");
  }

  // PG_ARGISNULL tests for SQL NULL,
  // but after the first pass we can have a
  // value that is non-SQL-NULL but still is C NULL.
  if (!PG_ARGISNULL(0)) {
    state = (VecArrayBuildState *)PG_GETARG_POINTER(0);
  }

  if (PG_ARGISNULL(1)) {
    // just return the current state unchanged (possibly still NULL)
    PG_RETURN_POINTER(state);
  }
  currentArray = PG_GETARG_ARRAYTYPE_P(1);
  if (ARR_NDIM(currentArray) == 0) {
    PG_RETURN_POINTER(state);
  }

  if (state == NULL) {
    // Since we have our first not-null argument
    // we can initialize the state to match its length.
    elemTypeId = ARR_ELEMTYPE(currentArray);
    if (elemTypeId != INT2OID &&
        elemTypeId != INT4OID &&
        elemTypeId != INT8OID &&
        elemTypeId != FLOAT4OID &&
        elemTypeId != FLOAT8OID) {
      ereport(ERROR, (errmsg("vec_to_mean input must be array of SMALLINT, INTEGER, BIGINT, REAL, or DOUBLE PRECISION")));
    }
    if (ARR_NDIM(currentArray) != 1) {
      ereport(ERROR, (errmsg("One-dimensional arrays are required")));
    }
    arrayLength = (ARR_DIMS(currentArray))[0];
    // Just start with all NULLs and let the comparisons below replace them:
    state = initVecArrayResultWithNulls(elemTypeId, FLOAT8OID, aggContext, arrayLength);
  } else {
    elemTypeId = state->inputElementType;
    arrayLength = state->state.nelems;
  }

  get_typlenbyvalalign(elemTypeId, &elemTypeWidth, &elemTypeByValue, &elemTypeAlignmentCode);
  deconstruct_array(currentArray, elemTypeId, elemTypeWidth, elemTypeByValue, elemTypeAlignmentCode,
      &currentVals, &currentNulls, &currentLength);
  if (currentLength != arrayLength) {
    ereport(ERROR, (errmsg("All arrays must be the same length, but we got %d vs %d", currentLength, arrayLength)));
  }

  for (i = 0; i < arrayLength; i++) {
    if (currentNulls[i]) {
      // do nothing: nulls can't change the result.
    } else if (state->state.dnulls[i]) {
      state->state.dnulls[i] = false;
      state->veccounts[i] = 1;
      switch (elemTypeId) {
        case INT2OID:   state->vecvalues[i].f8 = DatumGetInt16(currentVals[i]);  break;
        case INT4OID:   state->vecvalues[i].f8 = DatumGetInt32(currentVals[i]);  break;
        case INT8OID:   state->vecvalues[i].f8 = DatumGetInt64(currentVals[i]);  break;
        case FLOAT4OID: state->vecvalues[i].f8 = DatumGetFloat4(currentVals[i]); break;
        case FLOAT8OID: state->vecvalues[i].f8 = DatumGetFloat8(currentVals[i]); break;
        default: elog(ERROR, "Unknown elemTypeId!");
      }
    } else {
      state->veccounts[i] += 1;
      switch (elemTypeId) {
        case INT2OID:
          state->vecvalues[i].f8 += (DatumGetInt16(currentVals[i]) - state->vecvalues[i].f8) / state->veccounts[i];
          break;
        case INT4OID:
          state->vecvalues[i].f8 += (DatumGetInt32(currentVals[i]) - state->vecvalues[i].f8) / state->veccounts[i];
          break;
        case INT8OID:
          state->vecvalues[i].f8 += (DatumGetInt64(currentVals[i]) - state->vecvalues[i].f8) / state->veccounts[i];
          break;
        case FLOAT4OID:
          state->vecvalues[i].f8 += (DatumGetFloat4(currentVals[i]) - state->vecvalues[i].f8) / state->veccounts[i];
          break;
        case FLOAT8OID:
          state->vecvalues[i].f8 += (DatumGetFloat8(currentVals[i]) - state->vecvalues[i].f8) / state->veccounts[i];
          break;
        default:
          elog(ERROR, "Unknown elemTypeId!");
      }
    }
  }
  PG_RETURN_POINTER(state);
}

Datum vec_to_mean_finalfn(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(vec_to_mean_finalfn);

Datum
vec_to_mean_finalfn(PG_FUNCTION_ARGS)
{
  Datum result;
  VecArrayBuildState *state;
  int dims[1];
  int lbs[1];
  int i;

  Assert(AggCheckCallContext(fcinfo, NULL));

  state = PG_ARGISNULL(0) ? NULL : (VecArrayBuildState *)PG_GETARG_POINTER(0);

  if (state == NULL)
    PG_RETURN_NULL();

  // Convert from our pgnums to Datums:
  for (i = 0; i < state->state.nelems; i++) {
    if (state->state.dnulls[i]) continue;
    state->state.dvalues[i] = Float8GetDatum(state->vecvalues[i].f8);
  }

  dims[0] = state->state.nelems;
  lbs[0] = 1;

  result = makeMdArrayResult(&state->state, 1, dims, lbs, CurrentMemoryContext, false);
  PG_RETURN_DATUM(result);
}
