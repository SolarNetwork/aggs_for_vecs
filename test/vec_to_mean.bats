load test_helper

@test "int16 mean" {
  result="$(query "SELECT vec_to_mean(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{1,2,2.5}" ]
}

# @test "int32 mean" {
  # result="$(query "SELECT vec_to_mean('{1,1,5,2,0}'::integer[])")";
  # [ "$result" = "0" ]
# }

# @test "int64 mean" {
  # result="$(query "SELECT vec_to_mean('{1,1,5,2,0}'::bigint[])")";
  # [ "$result" = "0" ]
# }

# @test "float4 mean" {
  # result="$(query "SELECT vec_to_mean('{1,1,5,2,0}'::real[])")";
  # [ "$result" = "0" ]
# }

@test "float8 mean lots" {
  result="$(query "SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{1,2,2.5}" ]
}

@test "float8 mean none" {
  result="$(query "SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id = -1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "float8 mean one null" {
  result="$(query "SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id = 1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "float8 mean array of nulls" {
  result="$(query "SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id = 2")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

@test "float8 mean one not-null" {
  result="$(query "SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id = 4")";
  echo $result;
  [ "$result" = "{1,NULL,2}" ]
}

@test "float8 mean array of nulls and one other" {
  result="$(query "SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id IN (2, 4)")";
  echo $result;
  [ "$result" = "{1,NULL,2}" ]
}

@test "string mean" {
  run query "SELECT vec_to_mean(vals) FROM (VALUES (ARRAY['a']), (ARRAY['b'])) t(vals)"
  echo ${lines}
  [ "${lines[0]}" = "ERROR:  vec_to_mean input must be array of SMALLINT, INTEGER, BIGINT, REAL, DOUBLE PRECISION, or NUMERIC" ]
}

@test "numeric mean lots" {
  result="$(query "SELECT vec_to_mean(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{1.23,2.34,2.895}" ]
}

@test "numeric mean none" {
  result="$(query "SELECT vec_to_mean(nums) FROM measurements WHERE sensor_id = -1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "numeric mean one null" {
  result="$(query "SELECT vec_to_mean(nums) FROM measurements WHERE sensor_id = 1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "numeric mean array of nulls" {
  result="$(query "SELECT vec_to_mean(nums) FROM measurements WHERE sensor_id = 2")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

@test "numeric mean one not-null" {
  result="$(query "SELECT vec_to_mean(nums) FROM measurements WHERE sensor_id = 4")";
  echo $result;
  [ "$result" = "{1.23,NULL,2.34}" ]
}

@test "numeric mean array of nulls and one other" {
  result="$(query "SELECT vec_to_mean(nums) FROM measurements WHERE sensor_id IN (2, 4)")";
  echo $result;
  [ "$result" = "{1.23,NULL,2.34}" ]
}