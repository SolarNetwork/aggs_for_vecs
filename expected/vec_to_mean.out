-- smallint mean lots" {
SELECT vec_to_mean(smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_mean 
-------------
 {1,2,2.5}
(1 row)

-- smallint mean none" {
SELECT vec_to_mean(smallints) FROM measurements WHERE sensor_id = -1;
 vec_to_mean 
-------------
 
(1 row)

-- smallint mean one null" {
SELECT vec_to_mean(smallints) FROM measurements WHERE sensor_id = 1;
 vec_to_mean 
-------------
 
(1 row)

-- smallint mean array of nulls" {
SELECT vec_to_mean(smallints) FROM measurements WHERE sensor_id = 2;
   vec_to_mean    
------------------
 {NULL,NULL,NULL}
(1 row)

-- smallint mean one not-null" {
SELECT vec_to_mean(smallints) FROM measurements WHERE sensor_id = 4;
 vec_to_mean 
-------------
 {1,NULL,2}
(1 row)

-- smallint mean array of nulls and one other" {
SELECT vec_to_mean(smallints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_mean 
-------------
 {1,NULL,2}
(1 row)

-- int mean lots" {
SELECT vec_to_mean(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_mean 
-------------
 {1,2,2.5}
(1 row)

-- int mean none" {
SELECT vec_to_mean(ints) FROM measurements WHERE sensor_id = -1;
 vec_to_mean 
-------------
 
(1 row)

-- int mean one null" {
SELECT vec_to_mean(ints) FROM measurements WHERE sensor_id = 1;
 vec_to_mean 
-------------
 
(1 row)

-- int mean array of nulls" {
SELECT vec_to_mean(ints) FROM measurements WHERE sensor_id = 2;
   vec_to_mean    
------------------
 {NULL,NULL,NULL}
(1 row)

-- int mean one not-null" {
SELECT vec_to_mean(ints) FROM measurements WHERE sensor_id = 4;
 vec_to_mean 
-------------
 {1,NULL,2}
(1 row)

-- int mean array of nulls and one other" {
SELECT vec_to_mean(ints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_mean 
-------------
 {1,NULL,2}
(1 row)

-- bigint mean lots" {
SELECT vec_to_mean(bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_mean 
-------------
 {1,2,2.5}
(1 row)

-- bigint mean none" {
SELECT vec_to_mean(bigints) FROM measurements WHERE sensor_id = -1;
 vec_to_mean 
-------------
 
(1 row)

-- bigint mean one null" {
SELECT vec_to_mean(bigints) FROM measurements WHERE sensor_id = 1;
 vec_to_mean 
-------------
 
(1 row)

-- bigint mean array of nulls" {
SELECT vec_to_mean(bigints) FROM measurements WHERE sensor_id = 2;
   vec_to_mean    
------------------
 {NULL,NULL,NULL}
(1 row)

-- bigint mean one not-null" {
SELECT vec_to_mean(bigints) FROM measurements WHERE sensor_id = 4;
 vec_to_mean 
-------------
 {1,NULL,2}
(1 row)

-- bigint mean array of nulls and one other" {
SELECT vec_to_mean(bigints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_mean 
-------------
 {1,NULL,2}
(1 row)

-- real mean lots" {
SELECT vec_to_mean(reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_mean 
-------------
 {1,2,2.5}
(1 row)

-- real mean none" {
SELECT vec_to_mean(reals) FROM measurements WHERE sensor_id = -1;
 vec_to_mean 
-------------
 
(1 row)

-- real mean one null" {
SELECT vec_to_mean(reals) FROM measurements WHERE sensor_id = 1;
 vec_to_mean 
-------------
 
(1 row)

-- real mean array of nulls" {
SELECT vec_to_mean(reals) FROM measurements WHERE sensor_id = 2;
   vec_to_mean    
------------------
 {NULL,NULL,NULL}
(1 row)

-- real mean one not-null" {
SELECT vec_to_mean(reals) FROM measurements WHERE sensor_id = 4;
 vec_to_mean 
-------------
 {1,NULL,2}
(1 row)

-- real mean array of nulls and one other" {
SELECT vec_to_mean(reals) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_mean 
-------------
 {1,NULL,2}
(1 row)

-- float8 mean lots" {
SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_mean 
-------------
 {1,2,2.5}
(1 row)

-- float8 mean none" {
SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id = -1;
 vec_to_mean 
-------------
 
(1 row)

-- float8 mean one null" {
SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id = 1;
 vec_to_mean 
-------------
 
(1 row)

-- float8 mean array of nulls" {
SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id = 2;
   vec_to_mean    
------------------
 {NULL,NULL,NULL}
(1 row)

-- float8 mean one not-null" {
SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id = 4;
 vec_to_mean 
-------------
 {1,NULL,2}
(1 row)

-- float8 mean array of nulls and one other" {
SELECT vec_to_mean(floats) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_mean 
-------------
 {1,NULL,2}
(1 row)

